lib/plugin_helper.rb

```diff
@@ -0,0 +1,66 @@
+module RubyPluginHelper
+  # 'template' is the value of `target_mapping` from the inventory
+  # Returns an array of strings which are the lookups from the template
+  def config_lookup(template)
+    lookups = []
+    postwalk_vals(template) do |value|
+      # First collect the lookups to make
+      lookups << value if value.is_a?(String)
+      value
+    end
+    lookups.uniq.map { |lookup| lookup.split('.') }
+  end
+
+  # 'template' is the value of `target_mapping` from the inventory
+  # 'lookups' is an array of hashes, where each hash is the lookup for a target
+  # Returns an array of hashes to be sent to the inventory
+  def populate_mapping(template, lookups)
+    lookups.map do |lookup|
+      postwalk_vals(template) do |value|
+        if value.is_a?(String)
+          segments = value.split('.').map do |segment|
+            begin
+              # Turn it into an integer if we can
+              Integer(segment)
+            rescue ArgumentError
+              # Otherwise return the value
+              segment
+            end
+          end
+          answer = lookup.dig(*segments)
+          warn("Could not resolve lookup #{value} with #{lookup}") unless answer
+          answer
+        elsif value.is_a?(Array) || value.is_a?(Hash)
+          value
+        else
+          raise StandardError, 'Templated data must be a string, array, ' \
+            "or hash: #{value} is a #{value.class}"
```

Comment:
Maybe we can just say `must be string, array, etc, not a #{value.class}`. i'm not sure what could be in there, but it might not be cleanly printable.

Reply:
Hm yeah - this is mostly for ints or nils, my thinking being it'd be slightly easier to find where something was if it was printed. For nil it just looks weird though, and since we can't get the key that set it I'm not sure how helpful it is.

Reply:
ahh, yeah that is a good point. printing it will make it slightly easier to find.